File: examples/ruby/sanitizer_test/handlers.rb
# Ruby sanitizer-fixture — parallel handlers per sink family.
require 'cgi'
require 'shellwords'
require 'sqlite3'

class Handlers
  def initialize(db)
    @db = db
  end

  # --- Command injection --------------------------------------------------

  def cmd_raw(user_input)
    system("ping #{user_input}")
  end

  def cmd_safe(user_input)
    safe = Shellwords.escape(user_input)
    system("ping #{safe}")
  end

  # --- SQL injection ------------------------------------------------------

  def sql_raw(user_id)
    @db.execute("SELECT * FROM users WHERE id = '#{user_id}'")
  end

  def sql_safe(user_id)
    @db.execute("SELECT * FROM users WHERE id = ?", [user_id])
  end

  # --- XSS ----------------------------------------------------------------

  def xss_raw(name)
    "<p>Hello, #{name}</p>"
  end

  def xss_safe(name)
    safe = CGI.escapeHTML(name)
    "<p>Hello, #{safe}</p>"
  end
end

File: examples/perl/sanitizer_test/handlers.pl
# Perl sanitizer-fixture — parallel handlers per sink family.
use strict;
use warnings;
use HTML::Entities qw(encode_entities);
use URI::Escape    qw(uri_escape);
use String::ShellQuote;
use DBI;

# --- Command injection ---------------------------------------------------
sub cmd_raw {
    my ($input) = @_;
    return system("ping " . $input);
}

sub cmd_safe {
    my ($input) = @_;
    my $safe = shell_quote($input);
    return system("ping " . $safe);
}

# --- SQL injection -------------------------------------------------------
sub sql_raw {
    my ($dbh, $user_id) = @_;
    return $dbh->do("SELECT * FROM users WHERE id = '" . $user_id . "'");
}

sub sql_safe {
    my ($dbh, $user_id) = @_;
    my $sth = $dbh->prepare("SELECT * FROM users WHERE id = ?");
    $sth->execute($user_id);
    return $sth;
}

# --- XSS -----------------------------------------------------------------
sub xss_raw {
    my ($name) = @_;
    return "<p>Hello, " . $name . "</p>";
}

sub xss_safe {
    my ($name) = @_;
    my $safe = encode_entities($name);
    return "<p>Hello, " . $safe . "</p>";
}

# --- Open redirect -------------------------------------------------------
sub redirect_safe {
    my ($target) = @_;
    my $safe = uri_escape($target);
    return "/next?to=" . $safe;
}

1;
